In [2]:
import os
os.chdir('..') if os.getcwd().endswith('notebooks') else None

import pandas as pd
from sqlalchemy import create_engine
from dotenv import load_dotenv

load_dotenv()
DB_PASSWORD = os.getenv("DB_PASSWORD")
engine = create_engine(f"postgresql+psycopg2://postgres:{DB_PASSWORD}@localhost:5432/RetailPulse")

query = """
SELECT o.invoice, o.country, SUM(oi.quantity * oi.price) AS order_value
FROM order_items oi
JOIN orders o USING(invoice)
GROUP BY o.invoice, o.country
"""
orders = pd.read_sql(query, engine)
orders['is_uk'] = orders['country'] == 'United Kingdom'

orders.groupby('is_uk')['order_value'].describe()

,count,mean,std,min,25%,50%,75%,max
is_uk,,,,,,,,
False,2182.0,615.979743,1772.257998,-11880.84,15.0,320.32,732.8575,44051.60
True,23260.0,359.789456,982.999334,-25111.09,45.0,219.68,401.6625,49844.99


In [3]:
from scipy import stats

uk_values = orders[orders['is_uk'] == True]['order_value']
non_uk_values = orders[orders['is_uk'] == False]['order_value']

t_stat, p_value = stats.ttest_ind(uk_values, non_uk_values, equal_var=False)

print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.10f}")

t-statistic: -6.6571
p-value: 0.0000000000


In [4]:
print(f"p-value: {p_value:.2e}")

p-value: 3.48e-11


In [5]:
mean_diff = uk_values.mean() - non_uk_values.mean()
print(f"Mean difference (UK - Non-UK): ₹{mean_diff:.2f}")

# 95% confidence interval for the difference in means
ci = stats.ttest_ind(uk_values, non_uk_values, equal_var=False).confidence_interval(confidence_level=0.95)
print(f"95% CI for the difference: (₹{ci.low:.2f}, ₹{ci.high:.2f})")

Mean difference (UK - Non-UK): ₹-256.19
95% CI for the difference: (₹-331.66, ₹-180.72)


Finding: UK and non-UK customers show a statistically significant difference in average order value (Welch's t-test, t=-6.66, p=3.48×10⁻¹¹). Non-UK orders average ₹256 more than UK orders (95% CI: ₹181–₹332). This difference is highly unlikely to be due to random chance.

Caveat: This test establishes that a difference exists and its approximate size — it does not explain why. The cause may be order frequency, customer mix (retail vs. wholesale), or other factors not yet investigated.


In [6]:
result = stats.ttest_ind(uk_values, non_uk_values, equal_var=False)
print(f"Degrees of freedom: {result.df:.2f}")

Degrees of freedom: 2308.52


CHI- SQUARE TEST to determine whether country is associated with number of Cancelled order.

In [7]:
from scipy.stats import chi2_contingency
import numpy as np

contingency_table = np.array([
    [4005, 19255],   # UK: cancelled, not_cancelled
    [522, 1660]       # non-UK: cancelled, not_cancelled
])

chi2_stat, p_value, dof, expected = chi2_contingency(contingency_table)

print(f"Chi-square statistic: {chi2_stat:.4f}")
print(f"p-value: {p_value:.10f}")
print(f"Degrees of freedom: {dof}")
print(f"Expected counts (if H0 were true):\n{expected}")

Chi-square statistic: 60.8472
p-value: 0.0000000000
Degrees of freedom: 1
Expected counts (if H0 were true):
[[ 4138.74773996 19121.25226004]
 [  388.25226004  1793.74773996]]


In [8]:
n = contingency_table.sum()
cramers_v = np.sqrt(chi2_stat / (n * (min(contingency_table.shape) - 1)))
print(f"Cramér's V: {cramers_v:.4f}")

Cramér's V: 0.0489
